bpo-44347: [doc] clarify meaning of shutil.copytree's dirs_exist_ok kwarg

jdevries3133 opened the pull request:
* It now mirrors the documentation for os.makedirs
* Docstring in python source file was correspondingly updated

https://bugs.python.org/issue44347

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- issue-number: [bpo-44347](https://bugs.python.org/issue44347) -->
https://bugs.python.org/issue44347
<!-- /issue-number -->

jdevries3133 committed:
bpo-44347: [doc] clarify meaning of shutil.copytree's dirs_exist_ok kwarg

* It now mirrors the documentation for os.makedirs
* Docstring in python source file was correspondingly updated

jdevries3133 committed:
bpo-44347: [doc] clarify shutil.copytree dirs_exist_ok kwarg

* Add additional clarification: nonexistent parent directories will also
  be created.

jdevries3133 committed:
Merge branch 'main' of github.com:python/cpython into bpo-44347

nanjekyejoannah reviewed (commented):
Thanks for your contribution @jdevries3133, can you please fix the errors reported in the CI?

jdevries3133 committed:
bpo-44347: [doc|fix]: add NEWS.d entry, remove use of default context

jdevries3133 committed:
Merge branch 'main' of github.com:python/cpython into bpo-44347

akulakov commented on Doc/library/shutil.rst:
```diff
@@ -263,8 +261,13 @@ Directory and files operations
+   that supports the same signature (like :func:`~shutil.copy`) can be used.
+
+   If *dirs_exist_ok* is false (the default), a :exc:`FileExistsError` is
+   raised if *dst* already exists. Conversely, *dst* as well as any any parent
```
I think this phrasing is a bit unclear. 'conversely' implies that the following is opposite to how the default value works. In other words, I would read this sentence as saying that with dirs_exist_ok=False, dirs along the path to dst, as well as dst itself, will not be created.

jdevries3133 replied:
Yeah, you are totally correct; I've made it more confusing. I changed it to be much more explicit.

jdevries3133 committed:
bpo-44347: [doc] better document argument shutil.copytree(dirs_exist_ok=False)

* Be more explicit

jdevries3133 committed:
Merge branch 'main' of github.com:python/cpython into bpo-44347

jdevries3133 committed:
bpo-44347: [doc|fix]: add missed updated docstring from previous commit

zooba committed:
bpo-44381: Windows build now allows enabling control flow guard (GH-26645)

thsubaku9 committed:
bpo-40128: Fix IDLE autocomplete on macOS (GH-26672)

In particular, when running with tk8.6.8, as in PSF 3.9. 

Co-authored-by: Terry Jan Reedy <tjreedy@udel.edu>

mdickinson committed:
bpo-44339: Fix math.pow corner case to comply with IEEE 754 (GH-26606)

Change the behaviour of `math.pow(0.0, -math.inf)` and `math.pow(-0.0, -math.inf)` to return positive infinity instead of raising `ValueError`. This makes `math.pow` consistent with the built-in `pow` (and the `**` operator) for this particular special case, and brings the `math.pow` special-case handling into compliance with IEEE 754.

serhiy-storchaka committed:
bpo-43475: Fix the Python implementation of hash of Decimal NaN (GH-26679)

serhiy-storchaka committed:
Add more const modifiers. (GH-26691)

tim-one committed:
bpo-44376 - reduce pow() overhead for small exponents (GH-26662)

Greatly reduce pow() overhead for small exponents.

pablogsal committed:
bpo-44396: Update multi-line-start location when reallocating tokenizer buffers (GH-26676)



Automerge-Triggered-By: GH:pablogsal

Commit:
bpo-44389: Remove duplicate SSL_OP_NO_TLSv1_2 flag (GH-26680)

corona10 committed:
bpo-43425: Update test_c_parser not to use TempdirManager (GH-26693)

enjoy-binbin committed:
Fix typos in multiple files (GH-26689)

Co-authored-by: Terry Jan Reedy <tjreedy@udel.edu>

mdickinson committed:
Fix a potential reference-counting bug in long_pow (GH-26690)

tiran committed:
bpo-44389: Fix deprecation of OP_NO_TLSv1_3 (GH-26700)

Signed-off-by: Christian Heimes <christian@python.org>

rhettinger committed:
bpo-44310:  Note that lru_cache keep references to both arguments and results (GH-26715)

* Simplify the count_vowels example
* Hits and misses are fetched while a lock is held
* Add note that references are kept for arguments and return values
* Clarify behavior when *typed* is false.

markshannon committed:
bpo-44338: Port LOAD_GLOBAL to PEP 659 adaptive interpreter (GH-26638)

* Add specializations of LOAD_GLOBAL.

* Add more stats.

* Remove old opcache; it is no longer used.

* Add NEWS

markshannon committed:
Remove accidentally duplicated STAT_INC (GH-26718)

srittau committed:
bpo-38291: Remove mention of typing.io and typing.re again (GH-26113)

They were originally removed in GH-10173 per bpo-35089, but then
readded in GH-21574. Cf. bpo-38291 for decision to remove.

P403n1x87 committed:
Fix typo in lnotab_notes.txt (GH-26711)

Thanks for your contribution @P403n1x87

pablogsal committed:
bpo-44409: Fix error location in tokenizer errors that happen during initialization (GH-26712)

akulakov committed:
Fix a typo in _make_class_unpicklable() docstring (GH-26729)

markshannon committed:
Add extra stats for attribute misses (GH-26732)

Commit:
bpo-42064: Move sqlite3 types to global state (GH-26537)

* Move connection type to global state
* Move cursor type to global state
* Move prepare protocol type to global state
* Move row type to global state
* Move statement type to global state
* ADD_TYPE takes a pointer
* pysqlite_get_state is now static inline

vstinner committed:
bpo-42972: _thread.RLock implements the GH protocol (GH-26734)

The _thread.RLock type now fully implement the GC protocol: add a
traverse function and the Py_TPFLAGS_HAVE_GC flag.

vstinner committed:
bpo-44422: Fix threading.enumerate() reentrant call (GH-26727)

The threading.enumerate() function now uses a reentrant lock to
prevent a hang on reentrant call.

ethanfurman committed:
[Enum] improve test, add andrei kulakov to ACKS (GH-26726)

mdickinson committed:
bpo-43475: Add what's new entry for NaN hash changes (GH-26725)

ericsnowcurrently committed:
bpo-43693: Eliminate unused "fast locals". (gh-26587)

Currently, if an arg value escapes (into the closure for an inner function) we end up allocating two indices in the fast locals even though only one gets used.  Additionally, using the lower index would be better in some cases, such as with no-arg `super()`.  To address this, we update the compiler to fix the offsets so each variable only gets one "fast local".  As a consequence, now some cell offsets are interspersed with the locals (only when an arg escapes to an inner function).

https://bugs.python.org/issue43693

ethanfurman committed:
bpo-44342: [Enum] sync current docs to 3.10 (GH-26750)

encukou committed:
bpo-43795: Don't list private names in the limited API (GH-26740)

* Remove struct _node from the stable ABI list

This struct was removed along with the old parser in Python 3.9 (PEP 617)

* Stable ABI list: Use the public name "PyFrameObject" rather than "_frame"

* Ensure limited API doesn't contain private names

Names prefixed by an underscore are private by definition.

* Add a blurb

vstinner committed:
bpo-44422: threading.Thread reuses the _delete() method (GH-26741)

The _bootstrap_inner() method of threading.Thread now reuses its
_delete() method rather than accessing _active() directly. It became
possible since _active_limbo_lock became reentrant. Moreover, it no
longer ignores any exception when deleting the thread from the
_active dictionary.

sir-sigurd committed:
bpo-38211: Clean up type_init() (GH-16257)

1. Remove conditions already checked by assert()
2. Remove object_init() call that effectively creates an empty tuple and
checks that this tuple is empty

Fidget-Spinner committed:
bpo-44392: Add Py_GenericAlias to C API docs (GH-26724)

Also fix stable ABI type definitions

mdickinson committed:
bpo-44426: Use of 'complex' as a C variable name confuses Sphinx; change it to 'num'. (GH-26744)

j178 committed:
bpo-44389: Fix typo in ssl deprecation warning message (GH-26754)



`ssl.SSL_NO_TLS` should be `ssl.OP_NO_TLS`.

Commit:
bpo-43908: Make heap types converted during 3.10 alpha immutable (GH-26351)

* Make functools types immutable

* Multibyte codec types are now immutable

* pyexpat.xmlparser is now immutable

* array.arrayiterator is now immutable

* _thread types are now immutable

* _csv types are now immutable

* _queue.SimpleQueue is now immutable

* mmap.mmap is now immutable

* unicodedata.UCD is now immutable

* sqlite3 types are now immutable

* _lsprof.Profiler is now immutable

* _overlapped.Overlapped is now immutable

* _operator types are now immutable

* winapi__overlapped.Overlapped is now immutable

* _lzma types are now immutable

* _bz2 types are now immutable

* _dbm.dbm and _gdbm.gdbm are now immutable

markshannon committed:
Do not clear globals or builtins when calling clear() on a frame object. Reverts behavior to that of 3.10 and earlier. (GH-26768)

iritkatriel committed:
bpo-43024: improve signature (in help, etc) for functions taking sent… (GH-24331)



…inel defaults

rhettinger committed:
bpo-44310: Add a FAQ entry for caching method calls (GH-26731)

markshannon committed:
bpo-44032: Move pointer to code object from frame-object to frame specials array. (GH-26771)

brandtbucher committed:
Reorganize the pattern matching suite (GH-26661)

hroncok committed:
bpo-44451: Reset DeprecationWarning filters in test_importlib.test_entry_points_by_index (GH-26784)

This avoids the following error if DeprecationWarnings are ignored.

    ======================================================================
    ERROR: test_entry_points_by_index (test.test_importlib.test_metadata_api.APITests)
    Prior versions of Distribution.entry_points would return a
    ----------------------------------------------------------------------
    Traceback (most recent call last):
      File "/builddir/build/BUILD/Python-3.10.0b3/Lib/test/test_importlib/test_metadata_api.py", line 145, in test_entry_points_by_index
        expected = next(iter(caught))
    StopIteration
    ----------------------------------------------------------------------
    Ran 1402 tests in 2.125s
    FAILED (errors=1, skipped=18, expected failures=1)

ethanfurman committed:
bpo-43945: [Enum] reduce scope of new format() behavior (GH-26752)

* [Enum] reduce scope of new format behavior

Instead of treating all Enums the same for format(), only user mixed-in
enums will be affected.  In other words, IntEnum and IntFlag will not be
changing the format() behavior, due to the requirement that they be
drop-in replacements of existing integer constants.

If a user creates their own integer-based enum, then the new behavior
will apply:

    class Grades(int, Enum):
        A = 5
        B = 4
        C = 3
        D = 2
        F = 0

Now:  format(Grades.B)  -> DeprecationWarning and '4'
3.12:                   -> no warning, and 'B'

pablogsal committed:
Add a note about NameError/AttributeError suggestions with custom error functions (GH-26794)

rhettinger committed:
Remove dubious suggestion (GH-26789)

Remove the weakref example.  If a new instance is created and the same arguments are passes, it raises a ReferenceError.

tiran committed:
bpo-38820: Test with OpenSSL 3.0.0-beta1 (GH-26769)

Signed-off-by: Christian Heimes <christian@python.org>

zooba committed:
bpo-43298: Improved error message when building without the Windows SDK installed (GH-26800)

mdickinson committed:
bpo-44426: Fix use of the C keyword 'default' as a variable name (GH-26798)

terryjreedy committed:
Merge remote-tracking branch 'upstream/main' into pr_26643

srittau committed:
bpo-38291: DeprecationWarning when importing typing.{io,re} (#26719)

terryjreedy commented on Doc/library/shutil.rst:
```diff
@@ -230,9 +230,7 @@ Directory and files operations
-   named *dst* and return the destination directory. *dirs_exist_ok* dictates
-   whether to raise an exception in case *dst* or any missing parent directory
-   already exists.
+   named *dst* and return the destination directory.
```
The doc for dirs_exist_ok is misplaced here; moving it to the end us correct.

'or any missing parent directory' is wrong, so it is good that this is changed.  I think the name of the parameteris deceptive and was based on the mistaken understanding.  However, we cannot change it.  The corresponding parameter in os.makedirs is just 'exist_ok', referring to the target directory.

It is normal for parent directories to exist, so that should never raise.  But os.makedirs (plural) also makes intermediate dirs as needed.  This is the difference from os.makedir (singular).  So:
```suggestion
   named *dst* and return the destination directory.   Make all intermediate directories needed to contain *dst*.
```
Unlike os.makedirs, *dst* is not a leaf directory, because it is the root of the copied tree.  Hence the modification of the sentence from os.makedirs.

terryjreedy commented on Doc/library/shutil.rst:
```diff
@@ -263,8 +261,13 @@ Directory and files operations
+   If *dirs_exist_ok* is false (the default) and *dst* already exists, a
+   :exc:`FileExistsError` is raised. If *dirs_exist_ok* is true, *dst* as well
+   as any parent directory in the path to *dst* will be created if they do
+   not exist already.
```
The first sentence, copied from the os.makedirs doc, is correct if the latter is.  And this is all that os.makedirs says about the option.

The second sentence appears wrong in that intermediate are always created as needed.  That needs to be said at the top, as suggested in the previous comment.  I don't think more is needed because it should be clear enough that if
```suggestion
   If *dirs_exist_ok* is false (the default) and *dst* already exists, a
   :exc:`FileExistsError` is raised.
```

terryjreedy commented on Lib/shutil.py:
```diff
@@ -509,9 +509,6 @@ def copytree(src, dst, symlinks=False, ignore=None, copy_function=copy2,
-    dirs_exist_ok dictates whether to raise an exception in case dst or any
-    missing parent directory already exists.
-
     If exception(s) occur, an Error is raised with a list of reasons.
```
```suggestion
    Make all intermediate directories needed to contain *dst*.
    If exception(s) occur, an Error is raised with a list of reasons.
```

terryjreedy commented on Lib/shutil.py:
```diff
@@ -542,6 +539,10 @@ def copytree(src, dst, symlinks=False, ignore=None, copy_function=copy2,
+    If dirs_exist_ok is false (the default) and dst already exists, a
+    FileExistsError is raised. If dirs_exist_ok is true, dst as well as any
+    parent directory in the path to dst will be created if they do not exist
+    already.
```
```suggestion
    If dirs_exist_ok is false (the default) and dst already exists, a
    FileExistsError is raised.
```

jdevries3133 committed:
commit suggested change from @terryjreedy

Co-authored-by: Terry Jan Reedy <tjreedy@udel.edu>

jdevries3133 committed:
commit suggested change from @terryjreedy

Co-authored-by: Terry Jan Reedy <tjreedy@udel.edu>

jdevries3133 committed:
commit suggested change from @terryjreedy

Co-authored-by: Terry Jan Reedy <tjreedy@udel.edu>

jdevries3133 committed:
commit suggested change from @terryjreedy

Co-authored-by: Terry Jan Reedy <tjreedy@udel.edu>

jdevries3133 committed:
bpo-44347: fix spelling

terryjreedy committed:
Merge remote-tracking branch 'upstream/main' into pr_26643

akulakov commented on Misc/NEWS.d/next/Documentation/2020-06-10-16-55-55.bpo-44347.Fjvuw7.rst:
```diff
@@ -0,0 +1,2 @@
+Clarify the meaning of *dirs_exist_ok*, a kwarg passed to the shutil.copytree
```
Let's make it a link -- :func:`shutil.copytree`

jdevries3133 committed:
bpo-44347: add note to further clarify edge case

jdevries3133 committed:
bpo-44347: (small fix) remove redundant line from example

jdevries3133 committed:
bpo-44347: highlight relationship to os.makedirs

jdevries3133 committed:
small rewording fix

JelleZijlstra commented on Doc/library/shutil.rst:
```diff
@@ -263,8 +262,41 @@ Directory and files operations
+      Will create all those intermediary folders, even if *dirs_exist_ok* is
+      false.
+
+      However, the following will cause a :exc:`FileExistsError`, because it
```
Could this mention how the behavior depends on whether `dir_exist_ok` is passed? As written the example doesn't explicitly explain where it makes a difference.

JelleZijlstra replied:
Also, this example feels a bit excessive; we generally aim for concision in the docs. I see that lots of people have already contributed to this discussion though, so if they're in general agreement I'm OK with adding it.

jdevries3133 replied:
Yeah, I'm not crazy about the lack of brevity in it either, anymore. This was one of my first contributions and I was feeling myself a bit too much –– 

I will fix this up sometime soon when I get a chance. Thank you for the ping.

jdevries3133 committed:
Merge branch 'main' of github.com:python/cpython into bpo-44347

jdevries3133 committed:
revise old changes

* remove the note which is excessive
* remove all changes to `Lib/shutil.py`, pending approval of the new
  docs, I'll re-add revisions to shutil.copytree's docstring.
* revise first paragraph

JelleZijlstra commented on Doc/library/shutil.rst:
```diff
@@ -230,9 +230,9 @@ Directory and files operations
-   already exists.
+   named *dst* and return the destination directory.  All intermediate
+   directories needed to contain *dst* will also be created by default (see
+   kwarg *dirs_exist_ok*).
```
"kwarg" is a bit too informal. Maybe the parenthetical should just be removed?

jdevries3133 replied:
I'm ok with that, it's removed in c169b1d3d2d67e9be37a95b3843896b0bd3070d9

JelleZijlstra commented on Doc/library/shutil.rst:
```diff
@@ -263,8 +263,15 @@ Directory and files operations
+   operation will continue if it encounters existing directories, and files
+   within the *dst* tree will be overwritten by corresponding files from the
+   *src* tree. Internally, this parameter is passed directly to
+   :func:`os.makedirs`.
```
We should avoid making promises about the implementation. Maybe just remove this sentence.

jdevries3133 replied:
Ok, I removed that sentence in c169b1d3d2d67e9be37a95b3843896b0bd3070d9

JelleZijlstra reviewed (commented):
Thanks!

jdevries3133 committed:
apply @JelleZijlstra's feedback

JelleZijlstra reviewed (commented):
The text looks good! Can you update the docstring too and undraft the PR?

jdevries3133 committed:
update docstring
